src/liballoc/string.rs

```diff
@@ -1225,6 +1225,62 @@ impl String {
         ch
     }
 
+    /// Remove all matches of pattern `pat` in the `String`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(string_remove_matches)]
+    /// let mut s = String::from("Trees are not green, the sky is not blue.");
+    /// s.remove_matches("not ");
+    /// assert_eq!("Trees are green, the sky is blue.", s);
+    /// ```
+    ///
+    /// Matches will be detected and removed iteratively, so in cases where
+    /// patterns overlap, only the first pattern will be removed:
+    ///
+    /// ```
+    /// #![feature(string_remove_matches)]
+    /// let mut s = String::from("banana");
+    /// s.remove_matches("ana");
+    /// assert_eq!("bna", s);
+    /// ```
+    #[unstable(feature = "string_remove_matches", reason = "new API", issue = "72826")]
+    pub fn remove_matches<'a, P>(&'a mut self, pat: P)
+    where
+        P: for<'x> Pattern<'x>,
+    {
+        use core::str::pattern::Searcher;
+
+        let matches = {
+            let mut searcher = pat.into_searcher(self);
+            let mut matches = Vec::new();
+
+            while let Some(m) = searcher.next_match() {
+                matches.push(m);
+            }
+
+            matches
+        };
+
+        let len = self.len();
+        let mut shrunk_by = 0;
+
+        // SAFETY: start and end will be on utf8 byte boundaries per
+        // the Searcher docs
+        unsafe {
+            for (start, end) in matches {
```

Comment:
This is not the most efficient way to move the rejected substrings if there are a lot of matches. For example, consider `"1a2a3a4".to_string().remove_matches("a")`:
```
The sequence of intermediate strings is:
1a2a3a4 (the initial string)
12a3a4 (copy 2a3a4)
123a4 (copy 3a4)
1234 (copy 4)
```
The amount of copying work can grow quadratically with the input length, since a substring after the `n`th match will be moved `n` times.
A better solution is to keep a cursor to the end of the copied data and then to move each unmatched substring once:
```rust
let mut last_match_end = 0;
let mut written_end = 0;
for (start, end) in matches {
    // This copy can be skipped if the slice to copy is empty
    // or last_match_end == written_end
    self.vec.copy_within(last_match_end..start, written_end);
    last_match_end = end;
    written_end += start - last_match_end;
}
// This copy can be skipped if the slice to copy is empty
// or last_match_end == written_end
self.vec.copy_within(last_match_end.., written_end);
written_end += self.len() - last_match_end;
```
The code would be a bit simpler if you used `searcher.next_reject()` instead of `searcher.next_match()`, as those are the substrings that need to be copied.

Reply:
This is similar to the one I suggested, but yours use two variables here. https://github.com/rust-lang/rust/pull/71780#discussion_r484053735 I suggested to use a pointer instead (but maybe another variable may be needed too to not always copy to end).